src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,344 @@
+      compare: ''
+    };
+  }
+
+  public setFocus() {
+    if (this.autoCompleteRef && this.autoCompleteRef.current) {
+      this.autoCompleteRef.current.focus();
+    }
+  }
+
+  public onChange = (e: any) => {
+    const { suggestions, showSuggestions, userInput: previousUserInput, compare } = this.state;
+    const userInput = e.target.value;
+
+    this.props.contentChanged(userInput);
+
+    this.setState({
+      userInput
+    });
+
+    if (showSuggestions && suggestions.length) {
+      this.filterSuggestions(userInput, previousUserInput, compare, suggestions);
+    }
+    this.initialiseAutoComplete(userInput);
+  };
+
+  public onClick = (e: any) => {
+    const selected = e.currentTarget.innerText;
+    this.appendSuggestionToUrl(selected);
+  };
+
+  private initialiseAutoComplete = (url: string) => {
+    const lastCharacter = url.substring(url.length - 1);
+    switch (lastCharacter) {
+      case '/':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '?':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '=':
+
+        if (url.includes('?$')) {
+          this.getParameterEnums(url);
+        }
+
+        break;
+
+      case ',':
+        this.getParameterEnums(url);
+        break;
+
+      case '&':
+        this.getQueryParameters();
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  public onKeyDown = (e: any) => {
+    const { activeSuggestion, filteredSuggestions, showSuggestions } = this.state;
+
+    switch (e.keyCode) {
+      case KeyCodes.enter:
+        if (showSuggestions) {
+          const selected = filteredSuggestions[activeSuggestion];
+          this.appendSuggestionToUrl(selected);
+        } else {
+          this.props.runQuery();
+        }
+        break;
+
+      case KeyCodes.up:
+        if (activeSuggestion === 0) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion - 1 });
+        break;
+
+      case KeyCodes.down:
+        if (activeSuggestion - 1 === filteredSuggestions.length) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion + 1 });
+        break;
+
+      default:
+        break;
+    }
+
+  };
+
+  public displayLinkOptions = () => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (autoCompleteOptions) {
+      const suggestions: string[] = [];
+      const parameters = autoCompleteOptions.parameters;
+      if (parameters) {
+        const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+        if (parametersWithVerb) {
+          parametersWithVerb.links.forEach((value: string) => {
+            suggestions.push(value);
+          });
+
+          this.setState({
+            filteredSuggestions: suggestions,
+            suggestions,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
```

Comment:
please use "early return" for validation so that the code doesn't form arrowhead anti-pattern, e.g.:

```suggestion
    if (!autoCompleteOptions) {
      return;
    }
    const suggestions: string[] = [];
    const parameters = autoCompleteOptions.parameters;
    if (!parameters) {
      return;
    }
    const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
    if (!parametersWithVerb) {
      return;
    }
    parametersWithVerb.links.forEach((value: string) => {
      suggestions.push(value);
    });
    this.setState({
      filteredSuggestions: suggestions,
      suggestions,
      showSuggestions: true,
      compare: ''
    });
  }
}
```

Related discussion: https://stackoverflow.com/questions/17804005/how-to-prevent-the-arrowhead-anti-pattern